orbit/pkg/service/service_windows.go

```diff
@@ -0,0 +1,98 @@
+//go:build windows
+// +build windows
+
+package service
+
+import (
+	"os"
+	"time"
+
+	"github.com/fleetdm/fleet/v4/orbit/pkg/constant"
+	"github.com/fleetdm/fleet/v4/orbit/pkg/platform"
+
+	"github.com/rs/zerolog/log"
+	"golang.org/x/sys/windows/svc"
+)
+
+type windowsService struct {
+	rootDir string
+}
+
+func (m *windowsService) bestEffortShutdown(serviceRootDir string) {
+	err := platform.KillProcessByName(constant.DesktopAppExecName)
+	if err != nil {
+		log.Info().Err(err).Msg("The desktop app couldn't be killed")
+	}
+
+	if serviceRootDir != "" {
+		err = platform.KillFromPIDFile(serviceRootDir, constant.OsqueryPidfile, constant.OsquerydName)
+		if err != nil {
+			log.Info().Err(err).Msg("The osquery daemon process couldn't be killed")
+		}
+	}
+}
+
+func (m *windowsService) Execute(args []string, requests <-chan svc.ChangeRequest, changes chan<- svc.Status) (ssec bool, errno uint32) {
+	// Accepted service operations
+	const cmdsAccepted = svc.AcceptStop | svc.AcceptShutdown
+
+	// Expected service status update during initialization
+	changes <- svc.Status{State: svc.StartPending}
+	changes <- svc.Status{State: svc.Running, Accepts: cmdsAccepted}
+
+	// The listening loop below will keep listening for new SCM requests
+	for req := range requests {
+		switch req.Cmd {
+
+		case svc.Interrogate:
+			changes <- req.CurrentStatus
+
+		case svc.Stop, svc.Shutdown:
+
+			// Service shutdown was requested
+			// Updating the service state to indicate stop op started
+			changes <- svc.Status{State: svc.StopPending}
+
+			// Best effort tear down
+			m.bestEffortShutdown(m.rootDir)
+
+			// Updating the service state to indicate that stop finished
+			changes <- svc.Status{State: svc.Stopped}
+
+			// Dummy delay to allow the SCM to pick up the changes
+			time.Sleep(500 * time.Millisecond)
```

Comment:
just asking out of curiosity about the Windows ecosystem: is this a pseudo-random value? something you got by trial and error? a common practice?

Reply:
In this case it was kind of needed. The issue is that process is shutting down through os.Exit() and orbit need some time to send an RPC call to the SCM (async call in nature as SCM processing is not blocking). I've refactored the code to have this to be less dependent on the dummy sleep.

Reply:
thanks for the explanation!